ArmPkg/Drivers/SmmuDxe/IoMmu.c

```diff
@@ -0,0 +1,523 @@
+  IN BOOLEAN     SetFlagsOnly
+  )
+{
+  EFI_STATUS  Status;
+  UINT8       Level;
+  UINT64      Index;
+  PAGE_TABLE  *Current;
+
+  if (Root == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Current = Root;
+
+  // Traverse the page table to the leaf level
+  for (Level = 0; Level < PAGE_TABLE_DEPTH - 1; Level++) {
+    Index = PAGE_TABLE_INDEX (VA, Level);
+
+    if (Current->Entries[Index] == 0) {
+      PAGE_TABLE  *NewPage = (PAGE_TABLE *)AllocateAlignedPages (1, EFI_PAGE_SIZE);
+      if (NewPage == NULL) {
+        return EFI_OUT_OF_RESOURCES;
+      }
+
+      ZeroMem ((VOID *)NewPage, EFI_PAGE_SIZE);
+
+      Current->Entries[Index] = (PageTableEntry)(UINTN)NewPage;
+    }
+
+    if (!SetFlagsOnly) {
+      if (Valid) {
+        Current->Entries[Index] |= PAGE_TABLE_ENTRY_VALID_BIT; // valid entry
+      }
+    }
+
+    Status = UpdateFlags (Current, SetFlagsOnly, Flags, Index);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, " %a: UpdateFlags failed.\n", __func__));
+      return Status;
+    }
+
+    Current = (PAGE_TABLE *)((UINTN)Current->Entries[Index] & ~PAGE_TABLE_BLOCK_OFFSET);
+  }
+
+  // leaf level
+  if (Current != 0) {
+    Index = PAGE_TABLE_INDEX (VA, Level);
+
+    if (Valid && ((Current->Entries[Index] & PAGE_TABLE_ENTRY_VALID_BIT) != 0)) {
+      DEBUG ((DEBUG_INFO, "%a: Page already mapped\n", __func__));
+    }
+
+    if (!SetFlagsOnly) {
+      if (Valid) {
+        Current->Entries[Index]  = (PA & ~PAGE_TABLE_BLOCK_OFFSET); // Assign PA
+        Current->Entries[Index] |= PAGE_TABLE_ENTRY_VALID_BIT;      // valid entry
+      } else {
+        Current->Entries[Index] &= ~PAGE_TABLE_ENTRY_VALID_BIT; // only invalidate leaf entry
+      }
+    }
+
+    Status = UpdateFlags (Current, SetFlagsOnly, Flags, Index);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, " %a: UpdateFlags failed.\n", __func__));
+      return Status;
+    }
+  }
+
+  return Status;
+}
+
+/**
+  Update the page table with the given physical address and flags.
+
+  @param [in]  Root              Pointer to the root page table.
+  @param [in]  PhysicalAddress   Physical address to map.
+  @param [in]  Bytes             Number of bytes to map.
+  @param [in]  Flags             Flags to set for the mapping.
+  @param [in]  Valid             Boolean to indicate if the entry is valid.
+  @param [in]  SetFlagsOnly      Boolean to indicate if only flags should be set.
+
+  @retval EFI_SUCCESS            Success.
+  @retval EFI_INVALID_PARAMETER  Invalid parameter.
+  @retval EFI_OUT_OF_RESOURCES   Out of resources.
+**/
+EFI_STATUS
+EFIAPI
+UpdatePageTable (
+  IN PAGE_TABLE  *Root,
+  IN UINT64      PhysicalAddress,
+  IN UINT64      Bytes,
+  IN UINT64      Flags,
+  IN BOOLEAN     Valid,
+  IN BOOLEAN     SetFlagsOnly
+  )
+{
+  EFI_STATUS            Status;
+  EFI_PHYSICAL_ADDRESS  PhysicalAddressEnd;
+  EFI_PHYSICAL_ADDRESS  CurPhysicalAddress;
+
+  CurPhysicalAddress = ALIGN_DOWN_BY (PhysicalAddress, EFI_PAGE_SIZE);

```

Comment:
We cannot guarantee that the input PhysicalAddress or HostAddress passed into IoMmuMap is page aligned. When IoMmuAllocateBuffer is called, that buffer is page aligned, however, IoMmuMap() seems like it can be called with any address location in that previously allocated buffer. 

However, you are right, i should not be modifying the input PhysicalAddress, dropping the ALIGN_DOWN_BY. But still keeping the next line where I align up the end PhysicalAddressEnd, as that is needed.